Surface internal optimization modes (all rows vs first rows) at the SQL and configuration levels

dyemanov opened the pull request:
The idea is to force the optimizer to consider another (hopefully better) plan if only a subset or rows if fetched initially. For example, this is common for Delphi applications to load only the first records to the grid and continue fetching the following ones on demand (or in background). This feature may improve the response time in such applications. This was internally supported since v3 (used under the hood for FIRST <n> queries, and also for EXISTS subqueries), now these hints may be used for user queries as well.

Syntax `OPTIMIZE FOR {ALL | FIRST} ROWS` is taken from DB2 which also supports this feature. In fact DB2 allows to specify how many first rows should be taken into account by the optimizer, but I think it complicates its usage, so this is not supported by this patch. Although we could also add it if it's considered useful.

I'm not really happy with new conflicts in the parser, but making `OPTIMIZE` a reserved keyword is IMO not a good thing either.

dyemanov committed:
Surface internal optimization modes (all rows vs first rows) at the SQL and configuration levels.

livius2 commented:
Hi. 
Shouldn't it be settings by RDB$GET_CONTEXT? We can then turn it On before some statements and Off before another.

dyemanov commented:
It's kinda hint, but a very high-level one. It does not suggest the optimizer what exactly to do and you don't need to understand the plan under the hood. It only asks to use a different cost algorithm. As for RDB$CONTEXT, we never used them to turn something on or off. However, a session-level control statement may be useful for this feature to define per-attachment behaviour.

AlexPeshkoff committed:
Fixed #7691: 'with caller privileges' has no effect in triggers

ilya071294 committed:
Fix cases where the precedence relationship between a record page and a blob page is not set

1. While performing the intermediate garbage collection.
2. While modifying a blob field of a record.

hvlad committed:
Per-attachment cache of often used BDB's

hvlad committed:
Misc changes per @dyemanov requests

dyemanov committed:
Merge branch 'master' into optimize-for

dyemanov committed:
Add session-level control over the optimization strategy

asfernandes commented on src/dsql/StmtNodes.h:
```diff
@@ -1344,6 +1344,7 @@ class SelectNode final : public TypedNode<StmtNode, StmtNode::TYPE_SELECT>
 	bool dsqlForUpdate = false;
 	bool dsqlWithLock = false;
 	bool dsqlSkipLocked = false;
+	TriState dsqlOptimize;
```
It would look for informative some name like `dsqlOptimizeForFirst` as a bool here has no meaning.

dyemanov committed:
Better processing and optimization if IN <list> predicates (#7707)

* WIP

* Original (circa 2022) implementation of the IN LIST optimization, with some post-fixes and minor adjustments

* Make it possible to optimize IN <list> for middle segments in compund indices

* Avoid modifying the retrieval structure at runtime, it may be shared among concurrent requests

* Simplify the code a little. Better cost calculation. Support both root-based and sibling-based list scans inside the same plan node.

* Removed the unneeded const casts and other changed as suggested by Adriano

dyemanov committed:
Ensure the DDL trigger requests are cached (#7426)

While this commit does not fix all problems with cached triggers, I see no good reason to defer it.

dyemanov committed:
More informative name as suggested by Adriano

actions-user committed:
increment build number

dyemanov committed:
Postfix for replication server in Linux CS

dyemanov committed:
Protection from double deallocation of queued buffers, thanks to Vlad. Usually this should not happen, but possible after error thrown during commit (while journal/replica flushing).

actions-user committed:
increment build number

dyemanov committed:
Added missing (forgotten during refactoring) pre-sort before index scans. Fixed crash when all IN values are NULLs. Exclude duplicated IN values from the index scan.

AlexPeshkoff committed:
Fixed #7738: Crash on multiple connections/disconnections

actions-user committed:
increment build number

TreeHunter9 committed:
Fix wrong assert throw

It can be thrown by lastUsedPage(), when it trying to fetch incorrect page number of PIP

dyemanov committed:
Avoid tracing compilation of system triggers

AlexPeshkoff committed:
Misc

AlexPeshkoff committed:
Fixed most annoying bugs in pool<=>valgrind interaction - WIP

dyemanov committed:
Allow computable but non-invariant lists to be used for index lookup

actions-user committed:
increment build number

AlexPeshkoff committed:
Fixed compiling & running firebird under Valgrind control

actions-user committed:
increment build number

actions-user committed:
Misc.

actions-user committed:
increment build number

dyemanov committed:
Cost calculation adjustmnets for hash joins

dyemanov committed:
Better selectivity assumption for compound indices on empty tables

actions-user committed:
increment build number

dyemanov committed:
Reworked my yesterday's commit re. empty tables and indices

dyemanov committed:
Avoid storing garbage in idx_selectivity

dyemanov committed:
__func__ is now not only C99 but also a C++11 feature, should be supported by every modern compiler

hvlad committed:
This should fix blockage reported by Ilya Eremin (@ilya071294) privately.

dyemanov committed:
Fixed crash when IN predicate is delivered into aggregates/unions

dyemanov committed:
Cost-based approach for the first-rows optimization mode

dyemanov committed:
Merge branch 'master' into optimize-for

dyemanov committed:
Removed the weird duplication happened during merge
